### DISCUSSION TITLE
Make transports monitoring easier

### DISCUSSION THREAD

**Discussion by sam-mosleh** [encode/httpx#2580] [Ideas]
Currently, the only way to use public methods to monitor/instrument requests is to monkey-patch each transport class. By making `BaseTransport.handle_request` and `AsyncBaseTransport.handle_async_request` single entry points to all transport subclasses, it would be much easier to monitor them all (by patching only `BaseTransport` and `AsyncBaseTransport`), something like the below:

```python
async def _handle_async_request(
    self,
    request: Request,
) -> Response:
    """This method will be overridden by subclasses"""
    raise NotImplementedError("The '_handle_async_request' method must be implemented.")

async def handle_async_request(
    self,
    request: Request,
) -> Response:
    """This method will be the single entrypoint which would make monitoring easier by monkey-patching it"""
    return await self._handle_async_request(request)
```

I would be happy to try to open a PR for this. Thanks!

**Comment by lovelydinosaur** [comment]
> By making BaseTransport.handle_request and AsyncBaseTransport.handle_async_request single entry points to all transport subclasses

Can you explain to me what you mean here?

**Comment by sam-mosleh** [comment]
Right now, to monitor every request, we must create sub-classes for each transport:
```python
class MonitoredASGITransport(httpx.ASGITransport):
    async def handle_async_request(
        self,
        request: httpx.Request,
    ) -> httpx.Response:
        start_time = time.time()
        response = await super().handle_async_request(request)
        print("Request took", time.time() - start_time)
        return response
```
Since it is redundant for each transport class to override `handle_async_request`, it would be much easier to just monkey-patch `BaseTransport`s:
```python
async def monitored_handle_async_request(
    self,
    request: httpx.Request,
) -> httpx.Response:
    start_time = time.time()
    response = await old_handle(request)
    print("Request took", time.time() - start_time)
    return response


old_handle = httpx.AsyncBaseTransport.handle_async_request
httpx.AsyncBaseTransport.handle_async_request = monitored_handle_async_request
```
This is achieved by having `handle_async_request` as the entry point to all subclass's implementations.

**Comment by sam-mosleh** [comment]
@tomchristie By looking at the  [example commit](https://github.com/sam-mosleh/httpx/commit/2e6821ba45af2fa59989fe2717aaafc7302bfa60) I made, which is fully backward-compatible, you can review the entire change.
